packages/gateway/src/durable-objects/metrics.js

```diff
@@ -0,0 +1,150 @@
+ * @property {number} [averageResponseTime]
+ *
+ * @typedef {Object} ResponseStats
+ * @property {string} url gateway URL
+ * @property {boolean} ok request was successful
+ * @property {number} [responseTime] number of milliseconds to get response
+ * @property {boolean} [faster]
+ */
+
+/**
+ * Durable Object for keeping Metrics state.
+ */
+export class Metrics8 {
+  constructor(state, env) {
+    this.state = state
+    /** @type {Array<string>} */
+    this.ipfsGateways = JSON.parse(env.IPFS_GATEWAYS)
+
+    // `blockConcurrencyWhile()` ensures no requests are delivered until initialization completes.
+    this.state.blockConcurrencyWhile(async () => {
+      /** @type {Map<string, GatewayMetrics>} */
+      this.gatewayMetrics = new Map()
+
+      // Get state and initialize if not existing
+      const storedMetricsPerGw = await Promise.all(
+        this.ipfsGateways.map(async (gw) => {
+          /** @type {GatewayMetrics} */
+          const value = (await this.state.storage.get(gw)) || {
+            ...defaultCounter,
+          }
+
+          return {
+            gw,
+            value,
+          }
+        })
+      )
+
+      storedMetricsPerGw.forEach(({ gw, value }) => {
+        this.gatewayMetrics.set(gw, value)
+      })
+    })
+  }
+
+  // Handle HTTP requests from clients.
+  async fetch(request) {
+    // Apply requested action.
+    let url = new URL(request.url)
+    switch (url.pathname) {
+      case '/update':
+        const data = await request.json()
+        // Get updated Metrics
+        const updatedMetrics = this._getUpdatedMetrics(data)
+        // Save updated Metrics
+        await Promise.all(
+          this.ipfsGateways.map((gw) =>
+            this.state.storage.put(gw, updatedMetrics.get(gw))
+          )
+        )
+        return new Response()
+      case '/metrics':
+        const resp = {}
+        this.ipfsGateways.forEach((url) => {
+          resp[url] = this.gatewayMetrics.get(url)
+        })
+
+        return new Response(JSON.stringify(resp))
+      default:
+        return new Response('Not found', { status: 404 })
+    }
+  }
+
+  /**
+   * @param {ResponseStats[]} responseStats
+   */
+  _getUpdatedMetrics(responseStats) {
+    const updatedMetrics = this.gatewayMetrics
+
+    responseStats.forEach((stats) => {
+      const gwMetrics = {
+        ...updatedMetrics.get(stats.url),
+      }
+
+      if (!stats.ok) {
+        // Update request count
+        gwMetrics.totalReqCount += 1
+        gwMetrics.failedReqCount += 1
+
+        // Update metrics
+        updatedMetrics.set(stats.url, gwMetrics)
+        return updatedMetrics
+      }
+
+      // Update average response time
+      if (!gwMetrics.averageResponseTime) {
+        gwMetrics.averageResponseTime = stats.responseTime
+      } else {
+        gwMetrics.averageResponseTime = updateAverageValue(
+          stats.responseTime,
+          gwMetrics.averageResponseTime,
+          gwMetrics.totalReqCount
+        )
+      }
```

Comment:
The problem with reporting your own average response time is that it becomes less useful as time progresses as any changes to response time are less effectual on the value. We should have a rolling average, but I think this is something that grafana/prometheus should be able to calculate.

You might want to store the sum of request time instead for this: https://stackoverflow.com/questions/62428621/how-to-visualize-average-response-time-of-all-requests

Reply:
you are right, that makes a lot of sense! I did not thought well about it 🤦🏼‍♂️